uaclient/util.py

```diff
@@ -207,6 +202,40 @@ def subp(args: 'Sequence[str]', rcs: 'Optional[List[int]]' = None,
     return out.decode('utf-8'), err.decode('utf-8')
 
 
+def subp(args: 'Sequence[str]', rcs: 'Optional[List[int]]' = None,
+         capture: bool = False,
+         retry_sleeps: 'Optional[List[float]]' = None) -> 'Tuple[str, str]':
+    """Run a command and return a tuple of decoded stdout, stderr.
+
+     @param subp: A list of arguments to feed to subprocess.Popen
+     @param rcs: A list of allowed return_codes. If returncode not in rcs
+         raise a ProcessExecutionError.
+     @param capture: Boolean set True to log the command and response.
+     @param retry_sleeps: Optional list of sleep lengths to apply between
+        retries. Specifying a list of [0.5, 1] instructs subp to retry twice
+        on failure; sleeping half a second before the first retry and 1 second
+        before the next retry.
+
+    @return: Tuple of utf-8 decoded stdout, stderr
+    @raises ProcessExecutionError on invalid command or returncode not in rcs.
+    """
+    while True:
+        try:
+            out, err = _subp(args, rcs, capture)
+            break
+        except ProcessExecutionError as e:
+            message = str(e)
+            if retry_sleeps:
+                logging.debug(
+                    message + " Retrying %d more times.", len(retry_sleeps))
+                time.sleep(retry_sleeps.pop(0))
```

Comment:
This will mutate the list that's passed in, which would break consumer code that did something like:

```
sleeps = [1, 2, 3]
subp(..., retry_sleeps=sleeps)  # Would retry
subp(..., retry_sleeps=sleeps)  # Would not, because sleeps is now []
```